### PR TITLE
MainNoGUI: Shut down cleanly on SIGINT/SIGTERM

### PR DESCRIPTION
**Pull request by leoetlino**
This is the same as PR #3991, but for MainNoGUI.

nogui/headless will shut down cleanly on SIGINT and SIGTERM, just like it would when closing the render window.

The default signal handler will be restored after a first shutdown signal so a second signal will exit Dolphin forcefully.

Fixes [issue 9423](https://dolp.in/i9423).

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4243)

### CODE REVIEW

**Comment by lioncash** [comment]
*[Source/Core/DolphinWX/MainNoGUI.cpp, line 372 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4243#-KSXSsEIviI7-EKTtE7u:-KSXSsEIviI7-EKTtE7v:b-fgkx78) ([raw file](https://github.com/dolphin-emu/dolphin/blob/eaebb36e37d142af10435ef6fabbd52a0969fef7/Source/Core/DolphinWX/MainNoGUI.cpp#L372)):*
> ```cpp
> 
>   // Shut down cleanly on SIGINT and SIGTERM
>   struct sigaction sa;
> ```

`struct` isn't necessary.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4243)*

**Comment by leoetlino** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/DolphinWX/MainNoGUI.cpp, line 372 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4243#-KSXSsEIviI7-EKTtE7u:-KSXTEDh9ooZFQOOoM6d:bunv79y) ([raw file](https://github.com/dolphin-emu/dolphin/blob/eaebb36e37d142af10435ef6fabbd52a0969fef7/Source/Core/DolphinWX/MainNoGUI.cpp#L372)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `struct` isn't necessary.
</details>
I know it typically isn't, but it *is* here because there's a function named sigaction().

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4243)*

File: Source/Core/DolphinWX/MainNoGUI.cpp [hunk 1]
```diff
@@ -366,6 +368,21 @@ int main(int argc, char* argv[])
 
   platform->Init();
 
+  // Shut down cleanly on SIGINT and SIGTERM
+  struct sigaction sa;
+  sa.sa_handler = [](int unused) {
+    const char message[] = "A signal was received. A second signal will force Dolphin to stop.\n";
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-result"
+    write(STDERR_FILENO, message, sizeof(message));
```
Comment (delroth) [inline Source/Core/DolphinWX/MainNoGUI.cpp]: Can you just if (... < 0) { /* ignore */ } or something like that instead? Pragmas are ugly.

File: Source/Core/DolphinWX/MainNoGUI.cpp [hunk 2]
```diff
@@ -366,6 +368,21 @@ int main(int argc, char* argv[])
 
   platform->Init();
 
+  // Shut down cleanly on SIGINT and SIGTERM
+  struct sigaction sa;
+  sa.sa_handler = [](int unused) {
```
Comment (delroth) [inline Source/Core/DolphinWX/MainNoGUI.cpp]: Don't make that a lambda. It happens to work here but in practice I don't know if the standard specifies what you're doing. That lambda object should get "deleted" when "sa" leaves the scope.

Comment (leoetlino) [inline Source/Core/DolphinWX/MainNoGUI.cpp]: Fixed

**Pull request merged**